src/wallet/wallet2.cpp

```diff
@@ -1651,17 +1654,187 @@ void wallet2::scan_tx(const std::vector<crypto::hash> &txids)
     }
 
     for (auto& tx_info : res.txs)
-      txq.push(tx_info);
+    {
+      sorted_tx_entries.tx_entries.push_back(tx_info);
+      if (!tx_info.in_pool)
+      {
+        sorted_tx_entries.lowest_height = std::min(tx_info.block_height, sorted_tx_entries.lowest_height);
+        sorted_tx_entries.highest_height = std::max(tx_info.block_height, sorted_tx_entries.highest_height);
+      }
+    }
   }
 
+  // sort tx_entries by height from lowest height to highest; pool txs to the back
+  auto cmp_tx_entry = [](const cryptonote::COMMAND_RPC_GET_TRANSACTIONS::entry& l, const cryptonote::COMMAND_RPC_GET_TRANSACTIONS::entry& r)
+    { return l.block_height < r.block_height && !l.in_pool; };
+  std::sort(sorted_tx_entries.tx_entries.begin(), sorted_tx_entries.tx_entries.end(), cmp_tx_entry);
```

Comment:
This will only sort partially, if two of the txes are in the same block (or are both in the txpool). Do you care ?

Reply:
I don't see any code that relies on processing txs in the order they appear in a block (or in the pool), just by "height order." So this should suffice

Reply:
It could screw up cold signing if the m_transfers vector is ordered differently between hot and cold wallets (inputs are refered to by index IIRC). Possibly same for multisig.
All theoretical though. But having m_transfers shuffle itself a bit seems like the kind of bad idea that comes back to bite us later. I suppose it's nothing that can't be fixed later if it does come back though so not a show stopper I suppose.

Reply:
You're right, I missed that. Re-ordering txs like this can cause re-importing outputs from cold wallet to hot wallet to fail (I haven't checked pool/multisig behavior).

This is unfortunate. I don't see a nice way to handle this here without the daemon returning additional data that enables the client to sort txs in the same block (or client re-fetches entire blocks). Checking `output_indices` seems it would only work for RCT txs, not pre-RCT. Do you see a nice way to handle this?

Reply:
Multisig has the same issue, you were correct there too (`get_multisig_composite_key_image` relies on `m_transfers` to be in order).

For a 2/2 multisig, assume wallet A has scanned up to chain tip normally, and wallet B restores at a height > restore height and then uses `scan_tx([tx2, tx1])` to scan txs. If tx2 and tx1 are in the same block and wallet B calls `scan_tx` with the txs out of order, then wallet B will be borked.

Only way to prevent wallet B from getting borked here is for both wallets to order the txs in a consistent way, which goes back to my point above:

> I don't see a nice way to handle this here without the daemon returning additional data that enables the client to sort txs in the same block (or client re-fetches entire blocks)

Curious if you've got ideas here

Reply:
> so getting the blocks they're in does not leak any more information, and it should not be too much extra data.

Done in the latest